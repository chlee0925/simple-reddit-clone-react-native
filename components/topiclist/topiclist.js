import React from 'react';
import { Text, View, FlatList, Button } from 'react-native';
import { Topic } from '../topic/topic.js';

const numOfTopTopics = 20; // Number of top topics to show

// Sample topic data to populate initially
let sampleData = [
    { key: 1, topic: 'I am looking for good restaurants in town', upvote: 0, downvote: 0 },
    { key: 2, topic: 'Help! I just dropped my phone into the pool.', upvote: 2, downvote: 1 },
    { key: 3, topic: 'Best ways to bargain with sellers in Carousell', upvote: 1, downvote: 4 },
];
/**
 * Compare function to rank topics. Only 'upvote' property is considered as per requirements. 
 * Ranked in descending order.
 * @param {object} a Topic object 1 to compare
 * @param {object} b Topic object 2 to compare
 * @returns non-positive number if in order or positive number if out of order
 */
let upvoteCompareFn = function (a, b) { return b.upvote - a.upvote; };

/**
 * Screen displaying current top topics.
 */
export class TopicListScreen extends React.Component {
    constructor(props) {
        super(props);
        sampleData.sort(upvoteCompareFn);
        this.state = { 
            data: sampleData, 
            newKey: (sampleData.length + 1)
        }; // populate with sample data
    }

    /**
     * callback function for 'Topic' component to invoke when user presses 'upvote' button.
     * @param {number} index index of current topic in the data array.
     */
    upvote(index) {
        let data = this.state.data;
        data[index].upvote += 1;
        this._bubbleUpTopic(data, index);
        this.setState({ data: data });
    }

    /**
     * bubble up the topic in ranked data array. When a topic's upvote increases, this function
     * should be invoked to ensure the correct rank order of topics.
     * @param {object} data data array that holds topics
     * @param {number} index index of topic whose upvote has increased.
     */
    _bubbleUpTopic(data, index) {
        for (i = index; i > 0 && (upvoteCompareFn(data[i - 1], data[i]) > 0); i--) {                
            let temp = data[i - 1];
            data[i - 1] = data[i];
            data[i] = temp;
        }
    }

    /**
     * callback function for 'Topic' component to invoke when user presses 'downvote' button.
     * @param {number} index index of current topic in the data array.
     */
    downvote(index) {
        let data = this.state.data;
        data[index].downvote += 1;
        this.setState({ data: data });
    }

    /**
     * callback function for 'CreateTopicScreen' to invoke to create a new topic. New topic is appended
     * to the data array that holds topics.
     * @param {string} topic topic text user enters
     */
    createTopic(topic) {
        if (!topic) return;
        let data = this.state.data;
        let key = this.state.newKey;
        data.push({ key: key, topic: topic, upvote: 0, downvote: 0 });
        this.setState({ data: data, newKey: ++key })
    }

    render() {
        return (
            <View style={{ flex: 1 }}>
                <View style={{ marginBottom: 10 }}>
                    <Button
                        title='+ Create a new topic'
                        onPress={() => this.props.navigation.navigate('CreateTopic', {
                            createTopic: this.createTopic.bind(this)
                        })}
                    />
                </View>
                <FlatList
                    data={this.state.data.slice(0, numOfTopTopics)}
                    extraData={this.state}
                    renderItem={({ item, index }) =>
                        <Topic
                            index={index}
                            topic={item.topic}
                            upvote={item.upvote}
                            downvote={item.downvote}
                            upvoteFn={this.upvote.bind(this)}
                            downvoteFn={this.downvote.bind(this)}
                        />
                    }
                />
            </View>
        )
    }
}