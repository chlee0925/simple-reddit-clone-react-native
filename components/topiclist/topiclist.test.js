import React from 'react';
import { TopicListScreen } from './topiclist.js';

import renderer from 'react-test-renderer';

it('renders TopicListScreen Component', () => {
    const screen = renderer.create(
        <TopicListScreen />
    ).toJSON();
    expect(screen).toMatchSnapshot();
});