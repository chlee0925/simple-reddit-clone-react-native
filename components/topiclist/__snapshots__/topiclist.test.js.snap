// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TopicListScreen Component 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 10,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          + Create a new topic
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "downvote": 1,
          "key": 2,
          "topic": "Help! I just dropped my phone into the pool.",
          "upvote": 2,
        },
        Object {
          "downvote": 4,
          "key": 3,
          "topic": "Best ways to bargain with sellers in Carousell",
          "upvote": 1,
        },
        Object {
          "downvote": 0,
          "key": 1,
          "topic": "I am looking for good restaurants in town",
          "upvote": 0,
        },
      ]
    }
    disableVirtualization={false}
    extraData={
      Object {
        "data": Array [
          Object {
            "downvote": 1,
            "key": 2,
            "topic": "Help! I just dropped my phone into the pool.",
            "upvote": 2,
          },
          Object {
            "downvote": 4,
            "key": 3,
            "topic": "Best ways to bargain with sellers in Carousell",
            "upvote": 1,
          },
          Object {
            "downvote": 0,
            "key": 1,
            "topic": "I am looking for good restaurants in town",
            "upvote": 0,
          },
        ],
        "newKey": 4,
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#e3b5bc",
              "flex": 1,
              "marginBottom": 3,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            Topic: 
            Help! I just dropped my phone into the pool.
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginBottom": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              index: 
              0
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              upvote: 
              2
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              downvote: 
              1
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Up
                </Text>
              </View>
            </View>
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Down
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#e3b5bc",
              "flex": 1,
              "marginBottom": 3,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            Topic: 
            Best ways to bargain with sellers in Carousell
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginBottom": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              index: 
              1
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              upvote: 
              1
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              downvote: 
              4
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Up
                </Text>
              </View>
            </View>
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Down
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#e3b5bc",
              "flex": 1,
              "marginBottom": 3,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            Topic: 
            I am looking for good restaurants in town
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginBottom": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              index: 
              2
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              upvote: 
              0
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              downvote: 
              0
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Up
                </Text>
              </View>
            </View>
            <View
              accessibilityComponentType="button"
              accessibilityLabel={undefined}
              accessibilityTraits={
                Array [
                  "button",
                ]
              }
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={undefined}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Vote Down
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
